File: src/data_structure/linked_list/reverse.cpp
/*
  Reverse a linked list and return pointer to the head
  The input list will have at least one element  
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/
Node* Reverse(Node *head)
{
    Node *p, *q, *r;
    if (head == NULL || head->next == NULL)
    {
        return head;
    }
    p = head;
    q = head->next;
    p->next = NULL;
    do
    {
        r = q->next;
        q->next = p;
        p = q;
        q = r;
    } while(r != NULL);

    return p;
}

File: src/data_structure/linked_list/reverse.java
/*
  Insert Node at the end of a linked list 
  head pointer input could be NULL as well for empty list
  Node is defined as 
  class Node {
     int data;
     Node next;
  }
*/
    // This is a "method-only" submission. 
    // You only need to complete this method. 
Node Reverse(Node head) {
    Node p, q, r;
    if (head == null || head.next == null) {
        return head;
    }
    p = head;
    q = head.next;
    p.next = null;
    do {
        r = q.next;
        q.next = p;
        p = q;
        q = r;
    } while (r != null);
    return p;
}

